platform/pulp/plugin/changeset/main.py

```diff
@@ -0,0 +1,200 @@
+    """
+
+    def __init__(self, importer, adds=(), deletes=()):
+        """
+        Args:
+            importer (pulp.plugin.Importer): An importer.
+            adds (SizedIterable): The content to be added to the repository.
+            deletes (SizedIterable): The content IDs to be removed.
+
+        Notes:
+            The content to be added may already exist but not be associated
+            to the repository. Existing content is fetched and used instead
+            of provided content models as needed.
+        """
+        self.importer = importer
+        self.adds = adds
+        self.deletes = deletes
+        self.deferred = False
+
+    def _associate(self, content):
+        """
+        Associate content.
+
+        Args:
+            content (RemoteContent): The content to be associated.
+        """
+        log.debug(_('Associate content.'))
+        try:
+            association = RepositoryContent(
+                repository=self.importer.repository,
+                content=content)
+            association.save()
+        except IntegrityError:
+            # Duplicate
+            pass
+
+    def _disassociate(self, content):
+        """
+        (Un)associate content with the repository.
+
+        Args:
+            content (pulp.plugin.Content): A content model to be unassociated.
+        """
+        log.debug(_('Disassociate content.'))
+        RepositoryContent.objects.delete(
+            repository=self.importer.repository,
+            content=content)
+        for artifact in content.artifacts:
+            DownloadCatalog.objects.delete(
+                importer=self.importer,
+                artifact__in=artifact)
+
+    def _add_content(self):
+        """
+        Add wanted content.
+
+        Yields:
+            AddReport: A report for each content added.
+        """
+        activity = _('Adding content')
+        log.info(activity)
+        with Batch(DownloadIterator(self.adds, self.deferred)) as batch:
+            with ProgressBar(message=activity, total=len(self.adds)) as bar:
+                for plan in batch():
+                    artifact = plan.download.attachment
+                    artifact.model.downloaded = True
+                    content = artifact.content
+                    if not plan.succeeded:
+                        task = Task()
+                        task.append_non_fatal_error(plan.error)
+                        bar.increment()
+                        report = ChangeReport(ChangeReport.ADDED, content.model)
+                        report.errors.append(plan.error)
+                        yield report
+                        continue
+                    content.settle()
+                    if not content.settled:
+                        continue
+                    bar.increment()
+                    with transaction.atomic():
+                        self._update_catalog(content)
+                        self._associate(content.model)
```

Comment:
Still working on getting the DB transactions correct.

The goals for the current transaction alignment is:

1. Create a content (unit) and all its artifacts in 1 commit.
2. Add content to a repository and to the download catalog in 1 commit.

We _may_ be able to do all #2s in a single transaction.  This would also have performance advantages.  But I would need to play more with nested transactions to see how this could work with django ORM.   And, need to consider limitations of DB rollback buffer size. 

I think versioned repos will make this easier assuming the version created by a failed sync could be discarded.